### PR TITLE
Add support for chained matmuls

### PR DESCRIPTION
**Pull request by harsh-nod**
This PR adds support for chained matmuls. In
order to do this, we implement the following
- Support for op-specific vector_shapes
- Support for op-specific dim expansions (reduction
  dims of mmas will now always have chained expansions,
  even if they are not the induction variable of
  reduction ops)
- Adds a permute operator
- Adds anchor and vector_shapes to each op
- Adds reduction_dim to MMA ops

### CODE REVIEW

File: iree/turbine/kernel/wave/expansion.py [hunk 1]
```diff
@@ -187,8 +188,25 @@ def _expand_node(
     if (node, get_indexed_dims(dim_query, node), res_idx) in context:
         logger.debug(f"Already expanded node: {node} in {dim_query}")
         return context[(node, get_indexed_dims(dim_query, node), res_idx)]
+    elif isinstance(node, MMA):
+        # Handle expansion of MMA nodes whose reduction dim is not the same as the reduction
+        # dim of the parent reduction op.
```
Comment (harsh-nod) [inline iree/turbine/kernel/wave/expansion.py:193]: Handle the case where there is an mma without any reduction.

File: iree/turbine/kernel/wave/expansion.py [hunk 2]
```diff
@@ -230,15 +248,34 @@ def _expand_node(
...
-            new_arg = _expand_node(
-                arg,
-                trace,
-                restricted_dims,
-                dim_scaling,
-                context,
-                res_idx,
-            )
+        new_arg = None
+        if not isinstance(arg, Sequence):
```
Comment (harsh-nod) [inline iree/turbine/kernel/wave/expansion.py:252]: if not sequence arg = [arg]

File: iree/turbine/kernel/wave/expansion.py [hunk 3]
```diff
@@ -325,13 +367,107 @@ def _expand_reduction(
...
+    logger.debug(f"Expanding MMA reduction: {mma} in dims: {dim_query}")
+    idxc = IndexingContext.current()
+    for dim in mma.indexing_dims:
+        if dim not in dim_scaling and mma.vector_shapes[dim] > 0:
+            tile_size = idxc.get_static_value(dim)
+            dim_scaling[dim] = tile_size // mma.vector_shapes[dim]
+
+    expand_dims = set(mma.indexing_dims) - set([mma.reduction_dim])
+
+    # Store the original mma node and accumulator value for expansion.
```
Comment (harsh-nod) [inline iree/turbine/kernel/wave/expansion.py:411]: Explain this a little bit more as to why it is required.

File: iree/turbine/kernel/wave/expansion.py (hunk 1 above)
Comment (harsh-nod) [reply iree/turbine/kernel/wave/expansion.py:193]: done

File: iree/turbine/kernel/wave/expansion.py (hunk 2 above)
Comment (harsh-nod) [reply iree/turbine/kernel/wave/expansion.py:252]: done

File: iree/turbine/kernel/wave/expansion.py (hunk 3 above)
Comment (harsh-nod) [reply iree/turbine/kernel/wave/expansion.py:410]: done

File: iree/turbine/kernel/wave/index_sequence_analysis.py [hunk 4]
```diff
@@ -124,6 +129,7 @@ def set_node_indices(trace: CapturedTrace, constraints: list[Constraint]):
     mma_index, mma_slices = get_mma_dimensional_mapping(
         trace, get_hardware_constraint(constraints)
     )
+    trace.walk(partial(set_vector_shapes, constraints, mma_index, mma_slices))
```
Comment (raikonenfnu) [inline iree/turbine/kernel/wave/index_sequence_analysis.py:132]: No Action required: in the future perhaps we can do `set_vector_shapes` and `set_node_index` in a single walk. (still keeping the two functions apart though)

Comment (harsh-nod) [reply iree/turbine/kernel/wave/index_sequence_analysis.py:132]: good point. will keep that for a future PR.

File: iree/turbine/kernel/wave/utils.py [hunk 5]
```diff
@@ -219,18 +219,30 @@ def is_mma(node):
...
-        hardware_constraint.vector_shapes[m] = M
-        hardware_constraint.vector_shapes[n] = N
-        hardware_constraint.vector_shapes[k] = K
-
-    return mapping, capture_mma_slices([get_custom(x) for x in mma_nodes])
+        if custom not in mapping:
+            mapping[custom] = {}
+        mapping[custom][m] = 0
+        mapping[custom][n] = 1
+        mapping[custom][k] = 2
```
Comment (raikonenfnu) [inline iree/turbine/kernel/wave/utils.py:226]: No actions for now: Would be nice in the future to have an enum to represent [0, 1, 2] as [M, N, K] for better legibiltiy

Comment (harsh-nod) [reply iree/turbine/kernel/wave/utils.py:226]: makes sense, will add that as well.

**Pull request merged**